src/randomness/hash_chain.rs

```diff
@@ -0,0 +1,243 @@
+
+        for offset in
+            (0..num_full_blocks * KECCAK256_DIGEST_NUM_BYTES).step_by(KECCAK256_DIGEST_NUM_BYTES)
+        {
+            self.fill_random_bytes(
+                &mut random_bytes_out[offset..offset + KECCAK256_DIGEST_NUM_BYTES],
+            );
+        }
+
+        let num_tail_bytes = num_bytes % KECCAK256_DIGEST_NUM_BYTES;
+        if num_tail_bytes <= self.num_spare_bytes {
+            random_bytes_out[num_full_blocks * KECCAK256_DIGEST_NUM_BYTES..num_bytes]
+                .copy_from_slice(&self.spare_bytes[..num_tail_bytes]);
+            self.num_spare_bytes -= num_tail_bytes;
+
+            self.spare_bytes.copy_within(num_tail_bytes.., 0);
+        } else {
+            self.fill_random_bytes(
+                &mut random_bytes_out[num_full_blocks * KECCAK256_DIGEST_NUM_BYTES..num_bytes],
+            );
+        }
+    }
+
+    fn fill_random_bytes(&mut self, out: &mut [u8]) {
+        let num_bytes: usize = out.len();
+        assert!(
+            num_bytes <= KECCAK256_DIGEST_NUM_BYTES,
+            "Asked to get more bytes than one digest size"
+        );
+
+        let prandom_bytes = self.next_hash();
+        out.copy_from_slice(&prandom_bytes[..num_bytes]);
+
+        assert!(
+            self.num_spare_bytes < KECCAK256_DIGEST_NUM_BYTES + num_bytes,
+            "Not enough room in spare bytes buffer. Have {} bytes and want to add {} bytes",
+            self.num_spare_bytes,
+            KECCAK256_DIGEST_NUM_BYTES - num_bytes
+        );
+
+        self.spare_bytes
+            [self.num_spare_bytes..self.num_spare_bytes + (KECCAK256_DIGEST_NUM_BYTES - num_bytes)]
+            .copy_from_slice(&prandom_bytes[num_bytes..]);
+        self.num_spare_bytes += KECCAK256_DIGEST_NUM_BYTES - num_bytes;
+        self.counter += 1;
+    }
+
+    fn next_hash(&self) -> [u8; KECCAK256_DIGEST_NUM_BYTES] {
+        // TODO: below code is not efficient, but it works for now
+        let mut bytes_with_counter = [0u8; KECCAK256_DIGEST_NUM_BYTES * 2];
+        bytes_with_counter[..KECCAK256_DIGEST_NUM_BYTES].copy_from_slice(&self.digest);
+
+        // Copy the counter's serialized 64bit onto the MSB end of the buffer (stone-prover PR #875 decision).
+        // Serialized counter is in big-endian format.
+        bytes_with_counter[KECCAK256_DIGEST_NUM_BYTES * 2 - 8..]
+            .copy_from_slice(&self.counter.to_be_bytes());
+
+        let mut hasher = Keccak256::new();
+        hasher.update(bytes_with_counter);
+        let result = hasher.finalize();
+        let mut hash_bytes = [0u8; KECCAK256_DIGEST_NUM_BYTES];
+        hash_bytes.copy_from_slice(&result);
+        hash_bytes
+    }
+
+    pub fn update_hash_chain(&mut self, raw_bytes: &[u8]) {
+        let seed_increment: u64 = 0;
+        self.mix_seed_with_bytes(raw_bytes, seed_increment);
+    }
+
+    pub fn mix_seed_with_bytes(&mut self, raw_bytes: &[u8], seed_increment: u64) {
```

Comment:
Since it overwrites the entire state of the object, consider if it should actually take ownership of the object and return a new instance rather than have side-effects, e.g.


```rust
pub fn mix(self, bytes: &[u8], sep: u64) -> Self {
    let mut hasher = Keccak256::new();
    hasher.update(
        U256::from_be_bytes(self.digest)
            .add(U256::from(sep))
            .to_be_bytes(),
    );
    hasher.update(bytes);
    Self {
        digest: hasher.finalize().into(),
        ..Default::default()
    }
}
```

Or with the macro:

```rust
pub fn mix(self, bytes: &[u8], sep: u64) -> Self {
    Self {
        digest: keccak256!(
          U256::from_be_bytes(self.digest)
              .add(U256::from(sep))
              .to_be_bytes(),
           bytes
        ),
        ..Default::default()
    }
}
```